### PR TITLE
Enable reflection load ComImport assembly and Type.IsComObjectType

### LINKED ISSUE #16804: Fail to load type if the type has attribute to `ComImport` on Ubuntu/Mac/Docker container
_From @smadala on February 5, 2018 15:58_

### Description:
 Fail to load type if the type has attribute to `ComImport` on Ubuntu/Mac/Docker container.

### Repro Steps:
- git clone https://github.com/smadala/samples
- cd samples/ComImportIssue/consoleapp
- dotnet run

This works fine on windows.

### Expected Output:
```
Hello World!
nslib.ClassFactory
```
### Actual Output:
```
Hello World!

Unhandled Exception: System.TypeLoadException: Could not load type 'nslib.ClassFactory' from assembly 'nslib, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'.
   at consoleapp.Program.Test1()
   at consoleapp.Program.Main(String[] args) in /home/satya/Downloads/TestIssue/consoleapp/Program.cs:line 15
```
### Background:
In [vstest](https://github.com/Microsoft/vstest) we are using [`ComImport`](https://github.com/Microsoft/vstest/blob/347c59f73943921786bf9d5a7e42d2cf011ddd15/src/Microsoft.TestPlatform.ObjectModel/Navigation/NativeMethods.cs#L547
) to PInvoke [msdia](https://msdn.microsoft.com/en-us/library/t6tay6cz.aspx) API to get line numbers of functions for netstandard1.4. This causing issue like https://github.com/Microsoft/testfx/issues/341.

/cc: @mayankbansal018

_Copied from original issue: dotnet/corefx#26851_

### PR DESCRIPTION
**Pull request by luqunl**
Enable load ComImport assembly in xplat form and call Type.IsComObjectType. Most of these changes are related to move around #ifdef FEATURE_COMINTEROP.

Fix #16804

### CODE REVIEW

File: src/vm/method.hpp [hunk 1]
```diff
@@ -108,7 +108,6 @@ enum MethodClassification
     mcInstantiated = 5, // Instantiated generic methods, including descriptors
                         // for both shared and unshared code (see InstantiatedMethodDesc)
 
-#ifdef FEATURE_COMINTEROP
```
Comment (jkotas) [inline src/vm/method.hpp]: This ifdef should stay.

File: src/vm/class.h [hunk 2]
```diff
@@ -1958,11 +1953,11 @@ class EEClass // DO NOT CREATE A NEW EEClass USING NEW!
         VMFLAG_PREFER_ALIGN8                   = 0x00200000, // Would like to have 8-byte alignment
         // unused                              = 0x00400000,
 
-#ifdef FEATURE_COMINTEROP
         VMFLAG_SPARSE_FOR_COMINTEROP           = 0x00800000,
         // interfaces may have a coclass attribute
         VMFLAG_HASCOCLASSATTRIB                = 0x01000000,
```
Comment (jkotas) [inline src/vm/class.h]: Why did you need to pull in all these? I would think that `enum_flag_ComObject` should have been enough.

File: src/vm/methodtablebuilder.cpp [hunk 3]
```diff
@@ -1548,8 +1546,10 @@ MethodTableBuilder::BuildMethodTableThrowing(
         MethodTable* pMTParent = GetParentMethodTable();
```
Comment (jkotas) [inline src/vm/methodtablebuilder.cpp]: I do not think we need to bother will all COM interop specific validations without `FEATURE_COMINTEROP`. It will be incomplete anyway.

Just set `fIsComObjectType` bit and be done with it.

File: src/vm/class.h (hunk 2 above)
Comment (luqunl) [reply src/vm/class.h]: The methodtablebuilder will do some check for event interface during load time. I can pull  VMFLAG_HASCOCLASSATTRIB  out.

File: src/vm/method.hpp (hunk 1 above)
Comment (luqunl) [reply src/vm/method.hpp]: I guess somehow I need to know whether the method is com method or not.

File: src/vm/ecall.cpp [hunk 4]
```diff
@@ -333,8 +337,10 @@ PCODE ECall::GetFCallImpl(MethodDesc * pMD, BOOL * pfSharedOrDynamicFCallImpl /*
 
         // FCComCtor does not need to be in the fcall hashtable since it does not erect frame.
         return GetEEFuncEntryPoint(FCComCtor);
-    }
+#else
+    COMPlusThrow(kPlatformNotSupportedException, IDS_EE_ERROR_COM);
```
Comment (jkotas) [inline src/vm/ecall.cpp]: Nit: Indentation

File: src/vm/method.hpp (hunk 1 above)
Comment (jkotas) [reply src/vm/method.hpp]: Where is it that you need that?

File: src/vm/class.h (hunk 2 above)
Comment (jkotas) [reply src/vm/class.h]: All I can see is that these are just used to do some extra validation in methodtablebuilder. I do not think we need to bother with the extra COM-specific validation in methodtablebuilder with `FEATURE_COMINTEROP` disabled.

File: src/vm/ecall.cpp (hunk 4 above)
Comment (luqunl) [reply src/vm/ecall.cpp]: Fixed

File: src/vm/class.h [hunk 5]
```diff
@@ -1958,8 +1958,8 @@ class EEClass // DO NOT CREATE A NEW EEClass USING NEW!
         VMFLAG_PREFER_ALIGN8                   = 0x00200000, // Would like to have 8-byte alignment
         // unused                              = 0x00400000,
 
-#ifdef FEATURE_COMINTEROP
```
Comment (jkotas) [inline src/vm/class.h]: This change should not be needed.

Comment (luqunl) [reply src/vm/class.h]: Yes.. This one has been removed..

File: src/vm/method.hpp (hunk 1 above)
Comment (luqunl) [reply src/vm/method.hpp]: The problem is that in line https://github.com/dotnet/coreclr/blob/430331c011a0460670e2743ab08caa36f914f52d/src/vm/methodtablebuilder.cpp#L3079, I need to choose a method type: the ideal case is to choose correct one instead of FCall.

Comment (luqunl) [reply src/vm/method.hpp]: If I chose METHOD_TYPE_COMINTEROP, then I need to enable 
https://github.com/dotnet/coreclr/blob/32f0f9721afb584b4a14d69135bea7ddc129f755/src/vm/methodtablebuilder.inl#L516

Comment (jkotas) [reply src/vm/method.hpp]: What's wrong with choosing FCall?

Comment (luqunl) [reply src/vm/method.hpp]: that's why I remove #ifdef for mcComInterop

File: src/vm/class.h (hunk 2 above)
Comment (luqunl) [reply src/vm/class.h]: Okay. I am moving these out.

File: src/vm/method.hpp (hunk 1 above)
Comment (jkotas) [reply src/vm/method.hpp]: I think you should be able to ifdef the whole COM special case:

```
#ifdef FEATURE_COMINTEROP
                if (fIsClassComImport
                    || GetHalfBakedClass()->IsProjectedFromWinRT()
                    || bmtProp->fComEventItfType
                    )
                {
...

                }
                else
#endif // FEATURE_COMINTEROP
                if (dwMethodRVA == 0)
                {
                    type = METHOD_TYPE_FCALL;
....
```

File: src/vm/methodtablebuilder.cpp (hunk 3 above)
Comment (luqunl) [reply src/vm/methodtablebuilder.cpp]: Okay. working on

File: src/vm/method.hpp (hunk 1 above)
Comment (luqunl) [reply src/vm/method.hpp]: Then these methods will be treated as METHOD_TYPE_FCALL. I am afraid some reflection API will give wrong answers when checking methods in COM Types.

Comment (jkotas) [reply src/vm/method.hpp]: `MethodImplAttributes` are read directly from metadata. The method type does not influence them.

Comment (jkotas) [reply src/vm/method.hpp]: And if we find a place that does not deal with it correctly, we will fix it up.

Comment (luqunl) [reply src/vm/method.hpp]: Thanks a lot. just treat these methods as METHOD_TYPE_FCALL as suggested by you.

File: src/vm/interoputil.cpp [hunk 6]
```diff
@@ -1618,6 +1619,54 @@ BOOL CanCastComObject(OBJECTREF obj, MethodTable * pTargetMT)
...
+
+    if (!type.IsTypeDesc())
+    {
+        MethodTable *pMT = type.AsMethodTable();
+
+        if (pMT->IsComObjectType())
+        {
+            // May be __ComObject or typed RCW. __ComObject must have already been loaded
+            // if we see an MT marked like this so calling the *NoInit method is sufficient.
+#ifdef FEATURE_COMINTEROP
```
Comment (jkotas) [inline src/vm/interoputil.cpp]: The whole method is always going to return FALSE. You can make the ifdef wider.

Comment (luqunl) [reply src/vm/interoputil.cpp]: Fixed Update

**Comment by jkotas** [comment]
:shipit:
